memcopy.h

```diff
@@ -342,53 +291,108 @@ static inline unsigned char *byte_memset(unsigned char *out, unsigned len) {
         out += sz;
-        len -= 8;
+        len--;
     }
 
     return out;
 }
 
-/* Copy DIST bytes from OUT - DIST into OUT + DIST * k, for 0 <= k < LEN/DIST. Return OUT + LEN. */
-static inline unsigned char *chunk_memset(unsigned char *out, unsigned char *from, unsigned dist, unsigned len) {
-    if (dist >= len)
-        return chunk_memcpy(out, from, len);
-
-    Assert(len >= sizeof(uint64_t), "chunk_memset should be called on larger chunks");
-
-    /* Double up the size of the memset pattern until reaching the largest pattern of size less than SZ. */
-    unsigned sz = sizeof(uint64_t);
-    while (dist < len && dist < sz) {
-        copy_8_bytes(out, from);
+/*
+   Behave like memcpy, but assume that it's OK to overwrite at least
+   COPYCHUNKSIZE bytes of output even if the length is shorter than this, that
+   the length is non-zero, and that `from` lags `out` by at least COPYCHUNKSIZE
+   bytes (or that they don't overlap at all or simply that the distance is less
+   than the length of the copy).
 
-        out += dist;
-        len -= dist;
-        dist += dist;
+   Also assume that len > 0 on entry.
 
-        /* Make sure the next memcpy has at least SZ bytes to be copied.  */
-        if (len < sz)
-            /* Finish up byte by byte when there are not enough bytes left. */
-            return set_bytes(out, from, dist, len);
+   Aside from better memory bus utilisation, this means that short copies
+   (COPYCHUNKSIZE bytes or fewer) will fall straight through the loop without
+   iteration, which will hopefully make the branch prediction more reliable.
+ */
+static inline unsigned char *chunkcopy_relaxed(unsigned char *out, const unsigned char *from, unsigned len) {
+    --len;
+    storechunk(out, loadchunk(from));
+    out += (len % COPYCHUNKSIZE) + 1;
+    from += (len % COPYCHUNKSIZE) + 1;
+    len /= COPYCHUNKSIZE;
+#if defined(COPYCHUNK_UNROLL)
+    Assert(COPYCHUNK_UNROLL <= 8, "Unroll code needs more unrolling");
+    switch (len % COPYCHUNK_UNROLL) {
+    case 7: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 6: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 5: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 4: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 3: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 2: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 1: copychunk(&out, &from);
     }
-
-    return chunk_memcpy(out, from, len);
+    len /= COPYCHUNK_UNROLL;
+    while (len > 0) {
+        if (COPYCHUNK_UNROLL >= 1) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 2) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 3) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 4) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 5) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 6) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 7) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 8) copychunk(&out, &from);
```

Comment:
If you want to combine those two you're talking about Duff's device, which most people seem to regard as too esoteric to consider.